File: van_allen_beld.jl
using LinearAlgebra
using PyPlot
using OrdinaryDiffEq

function compute_magnetic_dipole_field(x, m)
    mu0 = 1.0
    r = norm(x)
    return mu0 / (4*pi) * (3 * dot(m, x) * x / r^5 - m / r^3)
end

function rhs(y, p, t)
    q, m, m_earth = p
    x, v = y[1:3], y[4:end]
    B = compute_magnetic_dipole_field(x, m_earth)
    a = q / m * cross(v, B)
    return vcat(v, a)
end

v_parallel_0 = 1e-8
v_orthogonal_0 = 1e-6
M = 10.0
m_earth = [0.0, 0.0, M]
q = 1.0
m = 1.0
L = 5.0
tmax = 1e7
x0 = [L, 0.0, 0.0,]
v0 = [0.0, v_orthogonal_0, v_parallel_0]

problem = ODEProblem(rhs, vcat(x0, v0), (0, tmax), (q, m, m_earth))
solver = Tsit5()
solution = solve(problem, solver, reltol=1e-6, abstol=1e-10)

figure()
a = 1
for i in 1:3
    for j in 1:3
        subplot(3, 3, a)
        plot(solution[i, 1:100:end], solution[j, 1:100:end], lw=1)
        xlabel(["x", "y", "z"][i])
        ylabel(["x", "y", "z"][j])
        global a += 1
    end
end
tight_layout()

figure()
plot3D(solution[1, 1:500], solution[2, 1:500], solution[3, 1:500])
xlabel("x")
ylabel("y")
zlabel("z")

show()

File: van_allen_beld.py
import numpy as np
import matplotlib.pyplot as plt
import numba
from scipy.integrate import solve_ivp
import time

@numba.njit
def compute_magnetic_dipole_field(x, m):
    mu0 = 1.0
    r = np.linalg.norm(x)
    return mu0 / (4*np.pi) * (3 * np.dot(m, x) * x / r**5 - m / r**3)

@numba.njit
def rhs(t, y, q_over_m, m_earth):
    x, v = y[:3], y[3:]
    B = compute_magnetic_dipole_field(x, m_earth)
    a = q_over_m * np.cross(v, B)
    dy = np.empty(6)
    dy[:3] = v
    dy[3:] = a
    return dy

v_parallel_0 = 1e-8
v_orthogonal_0 = 1e-6
M = 10.0
m_earth = np.array([0.0, 0.0, M])
q_over_m = 1.0
L = 5.0
tmax = 1e7
x0 = np.array([L, 0.0, 0.0])
v0 = np.array([0.0, v_orthogonal_0, v_parallel_0])

start = time.perf_counter()
sol = solve_ivp(rhs, (0, tmax), np.hstack([x0, v0]), args=(q_over_m, m_earth), rtol=1e-6, atol=1e-6)
assert sol.success
print("time:", time.perf_counter() - start, "seconds")

plt.figure()
plt.plot(sol.y[1], sol.y[2], lw=0.05)
plt.xlabel("y (long)")
plt.ylabel("z (lat)")
plt.title("van allen beld")
plt.show()

